stdlib/public/runtime/StaticBinaryELF.cpp

```diff
@@ -0,0 +1,316 @@
+//===-- StaticBinaryELF.cpp -------------------------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Parse a static ELF binary to implement lookupSymbol() address lookup.
+//
+//===----------------------------------------------------------------------===//
+
+#if defined(__ELF__) && defined(__linux__)
+
+#include "ImageInspection.h"
+#include "llvm/ADT/StringRef.h"
+#include <string>
+#include <vector>
+#include <elf.h>
+#include <fcntl.h>
+#include <pthread.h>
+#include <unistd.h>
+#include <linux/limits.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+using namespace std;
+using namespace llvm;
+
+#ifdef __LP64__
+#define ELFCLASS ELFCLASS64
+typedef Elf64_Ehdr Elf_Ehdr;
+typedef Elf64_Shdr Elf_Shdr;
+typedef Elf64_Phdr Elf_Phdr;
+typedef Elf64_Addr Elf_Addr;
+typedef Elf64_Word Elf_Word;
+typedef Elf64_Sym  Elf_Sym;
+typedef Elf64_Section Elf_Section;
+#define ELF_ST_TYPE(x) ELF64_ST_TYPE(x)
+#else
+#define ELFCLASS ELFCLASS32
+typedef Elf32_Ehdr Elf_Ehdr;
+typedef Elf32_Shdr Elf_Shdr;
+typedef Elf32_Phdr Elf_Phdr;
+typedef Elf32_Addr Elf_Addr;
+typedef Elf32_Word Elf_Word;
+typedef Elf32_Sym  Elf_Sym;
+typedef Elf32_Section Elf_Section;
+#define ELF_ST_TYPE(x) ELF32_ST_TYPE(x)
+#endif
+
+extern const Elf_Ehdr __ehdr_start;
```

Comment:
I actually went back to treating elfHeader as a ```struct Elf_Ehdr``` instead of a pointer and removed the read outside of the struct, using ```getauxval(AT_PHDR)``` instead.

As an aside, is there any option for ```clang++``` that warns of pointer misuse as above where the optimizer may make the assumptions that reads are not occuring outside of the defined object?